### DISCUSSION TITLE
Having a problem with live transcription cutting off/ending too soon.

### DISCUSSION THREAD

**Discussion by TheEvulKing** [deepgram/community#798] [General help] [answered]
Looking for any help with this issue. When I'm trying to speak a longer sentence or maybe pause slightly during the conversation. Deepgram will finish the conversation. I'm not sure if I need to extend the endpointing or utterance end time significantly more than I have it to fix the what I'm running into. Also just for clarity I'm live streaming from my microphone, not using any pre-recorded audio. 

```
is_finals = []

class TranscriptCollector:
    def __init__(self):
        self.reset()

    def reset(self):
        self.transcript_parts = []

    def add_part(self, part):
        self.transcript_parts.append(part)

    def get_full_transcript(self):
        return ' '.join(self.transcript_parts)

transcript_collector = TranscriptCollector()

def is_complete_sentence(text):
    nlp = spacy.load("en_core_web_sm")
    doc = nlp(text)
    
    if len(list(doc.sents)) > 0:
        return True
    return False
```

```
async def transcribe_audio(dg_client):
    try:
        dg_connection = dg_client.listen.live.v("1")
        utterance_ended = False
```

        def on_open(self, open, **kwargs):
            print(f"\n\n{open}\n\n")

        def on_message(self, result, **kwargs):
            global is_finals
            nonlocal transcript
            sentence = result.channel.alternatives[0].transcript
            if len(sentence):
                if is_complete_sentence(sentence):
                    print(f"speaker: {sentence}")
                    transcript = sentence
                else:
                    transcript += ' ' + sentence
            if result.is_final:
                is_finals.append(sentence)
                if result.speech_final and len(is_finals):
                    utterance = " ".join(is_finals)
                    print(f"Speech Final: {utterance}")
                    is_finals = []

                if not result.speech_final and not result.is_final and len(is_finals):
                   
                    transcript_collector.add_part(sentence)
                else:
                    print(f"Is Final: {sentence}")
            else:
                transcript_collector.add_part(sentence)
                full_sentence = transcript_collector.get_full_transcript()
                print(f"speaker: {full_sentence}")
                print(f"Interim Results: {sentence}")
                transcript_collector.reset()

        def on_metadata(self, metadata, **kwargs):
            print(f"Metadata: {metadata}")

        def on_speech_started(self, speech_started, **kwargs):
            print(f"\n\n{speech_started}\n\n")

        def on_utterance_end(self, utterance_end, **kwargs):
            print(f"\n\n{utterance_end}\n\n")
            global is_finals
            nonlocal utterance_ended
            if len(is_finals) > 10:
                utterance = " ".join(is_finals)
                print(f"\n\n{utterance_end}\n\n: {utterance}")
                is_finals = []
                utterance_ended = True

        def on_close(self, close, **kwargs):
            print(f"\n\n{close}\n\n")

        def on_error(self, error, **kwargs):
            print(f"Handled Error: {error}")

        def on_unhandled(self, unhandled, **kwargs):
            print(f"Unhandled Websocket Message: {unhandled}")

        dg_connection.on(LiveTranscriptionEvents.Open, on_open)
        dg_connection.on(LiveTranscriptionEvents.Transcript, on_message)
        dg_connection.on(LiveTranscriptionEvents.Metadata, on_metadata)
        dg_connection.on(LiveTranscriptionEvents.SpeechStarted, on_speech_started)
        dg_connection.on(LiveTranscriptionEvents.UtteranceEnd, on_utterance_end)
        dg_connection.on(LiveTranscriptionEvents.Close, on_close)
        dg_connection.on(LiveTranscriptionEvents.Error, on_error)
        dg_connection.on(LiveTranscriptionEvents.Unhandled, on_unhandled)

        options: LiveOptions = LiveOptions(
            model="nova-2",
            language="en-US",
            smart_format=True,
            encoding="linear16",
            channels=1,
            sample_rate=16000,
            interim_results=True,
            utterance_end_ms="3000",
            vad_events=True,
            endpointing=3000,
        )

        addons = {
            
            "no_delay": "true"
        }

        print("\n\nPress Enter to stop recording...\n\n")
        if dg_connection.start(options, addons=addons) is False:
            print("Failed to connect to Deepgram")
            return`

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by SandraRodgers** [comment, marked as answer]
The Deepgram endpointing algorithm relies on detecting silence in the audio. If you set `endpointing=2000`, it waits for 2000ms of silence to trigger a speech_final message. This works well with clear audio but is less effective with noisy signals, like phone calls, as it’s rare to have 2000ms of silence without some noise.

Longer endpointing times increase the likelihood of noise preventing the speech_final trigger. Shorter times (< 1000ms) work better, though it depends on the audio's noise level. By default, endpointing is set to 10ms, which is noise-resistant but may not suit cases where a clear end of speech is needed.

For times > 1000ms, the `utterance_end` feature is useful. It triggers based on detected silence between words in the transcript, not the audio signal. Deepgram transcriptions are produced every 3-5 seconds for `is_final` results, but with `interim_results` on, it produces transcripts every 1 second on average. So, using `utterance_end` with values > 1000ms can be effective.

**Reply by TheEvulKing** [reply]
Ok, thank you! I'll try that.

**Reply by TheEvulKing** [reply]
I made some adjustments based on everything you told me. Now everything seems to be working perfectly. Thank you very much for your help!

**Reply by lnomdedeu** [reply]
Hi @TheEvulKing , could you please share your working setup? Thanks
